### PR TITLE
Nano: Add ONNX quantize support for tensorflow

### PR DESCRIPTION
**Pull request by MeouSker77**
## Description

Add ONNX quantize support for Tensorflow

### 1. Why the change?

Support more features

### 2. User API changes

```python
from bigdl.nano.tf.keras import Model
model = Model(...)
dataset = Dataset(...)

# quantize directly
onnx_quantized_model = model.quantize(model, accelerator='onnxruntime', calib_dataset=train_dataset)
```

### 3. Summary of the change 

Add ONNX quantize support for Tensorflow and an UT

### 4. How to test?
- [ ] N/A
- [x] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### CODE REVIEW

File: python/nano/test/onnx/tf/test_inc_onnx.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+class TestONNX(TestCase):
+    def test_model_trace_onnx(self):
+        model = ResNet50(weights=None, input_shape=[224, 224, 3], classes=10)
+        model = Model(inputs=model.inputs, outputs=model.outputs)
+        input_examples = np.random.random((100, 224, 224, 3))
+        input_features = np.random.randint(0, 10, size=100)
+        
+        train_dataset = tf.data.Dataset.from_tensor_slices((input_examples,
+                                                            input_features)).batch(1)
```
Comment (rnwang04) [inline python/nano/test/onnx/tf/test_inc_onnx.py:32]: Here it seems calibration dataset for onnx+int8 requires `batch(1)`.
Does int8 quantization has this limit?
Maybe give more details for specific dataset requirements in the annotation of `calib_dataset: tf.data.Dataset`.

Other choice is do `batch(1)` within quantization.py and `unbatch` at last to align with int8.

**Review by rnwang04 — COMMENTED** [review]
Others LGTM.

File: python/nano/test/onnx/tf/test_inc_onnx.py (hunk 1 above)
Comment (MeouSker77) [reply python/nano/test/onnx/tf/test_inc_onnx.py:32]: > Other choice is do batch(1) within quantization.py and unbatch at last to align with int8.

good idea, take it.

File: python/nano/src/bigdl/nano/deps/neural_compressor/inc_api.py [hunk 2]
```diff
@@ -49,12 +49,14 @@ def quantize(model, dataloader=None, metric=None, **kwargs):
         from .pytorch.quantization import PytorchQuantization
         quantizer = PytorchQuantization(**not_none_kwargs)
     if 'onnx' in not_none_kwargs['framework']:
-        invalidInputError('torch' in str(type(dataloader)),
-                          errMsg="ONNXRuntime quantization only support in Pytorch.")
-        from .onnx.pytorch.quantization import PytorchONNXRuntimeQuantization
-        quantizer =\
-            PytorchONNXRuntimeQuantization(onnxruntime_session_options=onnxruntime_session_options,
-                                           **not_none_kwargs)
+        if 'torch' in str(type(dataloader)):
```
[the remark this answers is not in the document]
Comment (MeouSker77) [reply python/nano/src/bigdl/nano/deps/neural_compressor/inc_api.py]: fixed it by adding a new argument named `onnx_option`

File: python/nano/src/bigdl/nano/deps/neural_compressor/onnx/tensorflow/quantization.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        self.dataset = dataset
+        self.batch_size = 1     # it's necessary
+        self.dtype = dtype      # the dtype of dataset and model must be exactly the same
+
+    def __len__(self):
+        return len(self.dataset)
+
+    def __iter__(self):
+        for batch in self.dataset:
+            yield tuple(map(lambda x: x.numpy().astype(self.dtype), batch))
```
[the remark this answers is not in the document]
Comment (MeouSker77) [reply python/nano/src/bigdl/nano/deps/neural_compressor/onnx/tensorflow/quantization.py]: fixed it

File: python/nano/src/bigdl/nano/tf/keras/inference_utils.py [hunk 4]
```diff
@@ -82,6 +83,8 @@ def quantize(self,
                             Default: None, automatically get names from graph.
         :param outputs:     A list of output names.
                             Default: None, automatically get names from graph.
+        :param onnxruntime_session_options: The session option for onnxruntime, only valid when
```
[the remark this answers is not in the document]
Comment (MeouSker77) [reply python/nano/src/bigdl/nano/tf/keras/inference_utils.py:93]: updated it

File: python/nano/src/bigdl/nano/deps/neural_compressor/onnx/tensorflow/quantization.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        self.batch_size = 1     # it's necessary
+        self.dtype = dtype      # the dtype of dataset and model must be exactly the same
+
+    def __len__(self):
+        return len(self.dataset)
+
+    def __iter__(self):
+        for batch in self.dataset:
+            yield tuple(
+                map(lambda x: x.numpy().astype(self.dtype) if isinstance(x, tf.Tensor) else x,
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/deps/neural_compressor/onnx/tensorflow/quantization.py:66]: I think `x` could be a nested structure e.g. {"a": tf.Tensor, "b": (tf.Tensor, tf.Tensor), c: {"c1": tf.Tensor, "c2": tf.Tensor}}

Comment (yangw1234) [reply python/nano/src/bigdl/nano/deps/neural_compressor/onnx/tensorflow/quantization.py:66]: shouldn't we also process the x in the else branch?

Comment (MeouSker77) [reply python/nano/src/bigdl/nano/deps/neural_compressor/onnx/tensorflow/quantization.py:66]: fix it in #6804

**Pull request merged**